modules/programs/afew.nix

```diff
@@ -6,21 +6,134 @@ let
 
   cfg = config.programs.afew;
 
+  dag = config.lib.dag;
+
+  mkIniKeyValue = key: value:
+    let
+      tweakVal = v:
+        if isString v then v
+        else if isList v then concatMapStringsSep ";" tweakVal v
+        else if isBool v then (if v then "true" else "false")
+        else toString v;
+    in
+      "${key}=${tweakVal value}";
+
+  mkDag = filters:
+    let
+      mkNode = nodeName: filter:
+        let
+          before = (filter.before or cfg.defaultBefore);
+          after = (filter.after or cfg.defaultAfter);
+          filterName = (filter.class or "Filter");
+          filterKV = (removeAttrs filter ["after" "before" "class"]) // {
+            "#nodeName" = nodeName;
+            "#before" = before;
+            "#after" = after;
+          };
+        in
+          dag.entryBetween before after (nameValuePair filterName filterKV);
+    in
+      mapAttrs mkNode filters;
+
+  /* Return a topologically sorted list of DAG entries encapsulating
+     name value pairs.
+  */
+  reorderFilters = filters:
+    (dag.topoSort (mkDag filters)).result;
+
+  /* Convert a topologically sorted list of DAG nodes to a list of INI
+     sections.  The section name is appended with an index number to
+     prevent clashes, as is required by afew.
+  */
+  toOrderedINI = {
+    mkKeyValue ? mkKeyValueDefault {} "="
+  }: filters:
+    let
+      mkSection = iota: node: ''
+        [${node.data.name}.${toString iota}]
+        ${toKV node.data.value}
+      '';
+      toKV = attrs: generators.toKeyValue { inherit mkKeyValue; } attrs;
+    in
+      concatStringsSep "\n" (imap0 mkSection filters);
+
 in
 
 {
   options.programs.afew = {
     enable = mkEnableOption "the afew initial tagging script for Notmuch";
 
+    defaultBefore = mkOption {
```

Comment:
Are the `defaultBefore`, `defaultAfter`, and `defaultFilters` options necessary? I figure these could simply be hard coded?

Reply:
I will hardcode them for now to simplify the current implementation, and we'll see if anybody complains about the lack of flexibility.